### ISSUE TITLE
Argument to control and reporting of DF and CI

### ISSUE THREAD

**Issue by strengejacke** [easystats/parameters#570]
I try to integrate the related issues here.

- [ ] Inconsistency in reporting interval type (#537, #523)
- [ ] Argument to control type of CI or DF (#559)

## Inconsistency in reporting interval type

> I think it would be good to just always report the type of interval (not only for "bayesian" models) and to use a simpler language:

`Uncertainty intervals are <type> <method> intervals.`

## Argument to control type of CI or DF

> Personally, I don't think we need two arguments, but the argument should be called 'ci_method'. 
> But isn't not only used for CI, but also for the p-value... :/

**Comment by strengejacke** [comment]
Some questions:

1. If we report type of intervals, which style? Suggestion from @bwiernik:

`Uncertainty intervals are <type> <method> intervals.`

```
Uncertainty intervals are equal-tailed Wald intervals.
Uncertainty intervals are equal-tailed bootstrap intervals.
Uncertainty intervals are equal-tailed parametric bootstrap intervals.
Uncertainty intervals are equal-tailed MCMC intervals.
Uncertainty intervals are bias-corrected accelerated bootstrap intervals.
Uncertainty intervals are highest-density MCMC intervals.
Uncertainty intervals are highest-density bootstrap intervals.
```


2. If we report type of interval, should we also report DF? If yes, how to combine?

`Uncertainty intervals are <type> <method> intervals, based on <xy> approximated degrees of freedom.`
or
`Uncertainty intervals are <type> <method> intervals and based on <distribution> approximation`


3. CIs are different depending on how the DFs are approximated (Wald-z, residual DF, Kenward-Roger, Satterthwaite, ...). In these cases, it is enough to know the DF-method to know how CIs were calculated. We have the argument `df_method` to do this. Advantage: p-values are also based on DF, so `df_method` refers to CI and p. However, sometimes we have profiled or bootstrapped CIs. Now the "DF method" is different from the "CI method", e.g. profiled CIs, but Wald-Z (Inf) DF, or profiled CIs and residual DF for the t-statistic. Here we would need to different arguments?

**Comment by mattansb** [comment]
2. Yes, I think we should report both, and
3. I think that there should be two arguments (Also, don't we support some bootstrap based p-values?)

**Comment by IndrajeetPatil** [comment]
1. I like Brenton's suggestion. Unless someone else has a suggestion for a template, we can stick to this one.

2. I prefer this option to also report `df` method.
    > Uncertainty intervals are <type> <method> intervals, based on <xy> approximated degrees of freedom.

3. I agree with Mattan, we can have two arguments: `df_method` and `ci_method`, with the defaults being 
   `df_method = xxx, ci_method = df_method`. Those who will be changing the defaults for these arguments can be assumed 
    to know what they are doing, so there shouldn't be any confusion for anyone by the introduction of another argument.
   Additionally, we also have a precedent for `ci_method` argument, which is used in `bayestestR` in a completely different 
   context, so we can expect some `easystats` users to be familiar with the presence and usage of this argument.

**Comment by bwiernik** [comment]
I don't like the idea of two arguments. Given precedence of `ci_method` in bayestestR and similar arguments in broom and many packages' `confint()` functions, I expect a lot of users to intuitively think of that one. That makes it pretty likely that p values and confidence intervals will be based on different methods accidentally. 

I don't think it's that big a deal to call the argument `ci_method()` and note in the documentation (and maybe table footnote if not too verbose) that p values use the same method.

Perhaps:

`Uncertainty intervals (equi-tailed) and p values (two-tailed) computed using normal distribution approximation`

If it's really bothersome to use just `ci_method` and we have both, `ci_method` should override `df_method` if it is set.

**Comment by mattansb** [comment]
> If it's really bothersome to use just ci_method and we have both, ci_method should override df_method if it is set.

Yup - that seems fine (doesn't matter to me which defaults to which) and reasonable given design elsewhere in easystats/broom.

**Comment by strengejacke** [comment]
This would mean, we may have, e.g.:

- `model_parameters(model, ci_method = "hdi")` (which works only for Bayesian models)
- `model_parameters(model, ci_method = "wald")` (which works only(?) for freq. models)
- `model_parameters(model, ci_method = "satterthwaite")` (which looks strange, because "satterthwaite" is no CI-method per se, and it works only for freq. mixed models)
- `model_parameters(model, ci_method = "profile")` (which works only for freq. mixed models and glm, and maybe some others, but p-values are still based on Inf/residual DF)
- ...

Although I prefer having one argument, this looks a bit strange to me...

**Comment by mattansb** [comment]
That's why it makes more sense to have `df_method` control `ci_method` be default:

- `model_parameters(model, ci_method = "hdi")` (which works only for Bayesian models)
- `model_parameters(model, df_method = "wald")`
- `model_parameters(model, df_method = "satterthwaite")` for mixed freq
- `model_parameters(model, df_method = "satterthwaite", ci_method = "profile")` for mixed freq with CI for random effects

**Comment by bwiernik** [comment]
The one thing I really don't want is for there to be two separate arguments that act independently or something like this will happen:

`parameters(mod, df_method = "normal", ci_method = "boot")`

For profile CIs, we really out to use LRT or profile p values. 

("hdi" conceptually works for any estimator, just in most cases it's the same as the equi-tailed because symmetrical sampling dists)

**Comment by bwiernik** [comment]
`model_parameters(model, df_method = "satterthwaite", ci_method = "profile")` 

This doesn't make any sense to me. These aren't compatible forms of inference.

**Comment by mattansb** [comment]
Yeah, I see your point 🎯

**Comment by strengejacke** [comment]
> `model_parameters(model, df_method = "satterthwaite", ci_method = "profile")`
> 
> This doesn't make any sense to me. These aren't compatible forms of inference.

![your-reply-while-useful-does-not-answer-the-question](https://user-images.githubusercontent.com/26301769/127566683-ae3beffd-e080-4769-bdf5-bbafa4be5e8a.jpg)

**Comment by bwiernik** [comment]
I mean that satterthwaite DF and likelihood-based inference make different assumptions about the nature of the model and its sampling distribution. They don't necessarily yield the same inference (eg, if you computed a p value via satterthwaite and CI by profile, whether p < .05 and 95% CI exclude 0 won't necessarily agree). P values and CIs are just two different summaries of the same uncertainty distribution, so it doesn't make much sense to me to compute them using two different distributions.

Given that, I don't know that we need to allow users to request two different methods, and like I said above, I expect it would usually be accidental. If we want to allow users to do so, that should require an explicit argument saying it's intentional.

**Comment by mattansb** [comment]
I am convinced - the arg should be `ci_method`. With:

| Model | Available `ci_method`s | What happens to p-values? |
| -- | -- | -- |
| **Bayesian models** | HDI, ETI, SI | N/A |
| **Freq LMM** | Wald (default), ml1, betwithin, Satterthwaite, Kenward, uniroot, profile | Match the method, where for the latter two (?), the p-value is a Wald p-value |
| **Freq GLM(M)** | Wald (default), profile  | Both give Wald p-values |
| **`bootstrap = TRUE`** | HDI, ETI, and all the boot methods (bca, bci, quantile, ...) | p-values are based on the bootstrapped dist. (assuming translation equivariance, which is meh) |

(The various robust CIs / P-values can also be added here.)

**Comment by bwiernik** [comment]
Yes, it would he great to put the various Huber-White, etc methods here. Where do we call those currently

**Comment by strengejacke** [comment]
You mean robust estimation? With `robust = TRUE`, it calls one of https://github.com/easystats/parameters/blob/main/R/robust_estimation.R

**Comment by strengejacke** [comment]
One thing that I'm not 100% sure about is whether the different approx. method should only change DF, or also SE and test-statistic? *lmerTest* e.g. changes SE (and thus, t-statistic) and DF:

``` r
library(lmerTest)
m3 <- lmerTest::lmer(Petal.Length ~ Sepal.Length + (1 | Species), data = iris)
coef(summary(m3, ddf = "lme4"))
#>               Estimate Std. Error     t value
#> (Intercept)  0.0525246 0.95576245  0.05495571
#> Sepal.Length 0.6341373 0.04524917 14.01433934
coef(summary(m3, ddf = "Kenward-Roger"))
#>               Estimate Std. Error         df     t value     Pr(>|t|)
#> (Intercept)  0.0525246 0.95583828   2.342223  0.05495135 9.605137e-01
#> Sepal.Length 0.6341373 0.04529607 146.288076 13.99983109 8.598429e-29
```

<sup>Created on 2021-08-04 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by mattansb** [comment]
That's only for the Kenward-Roger method:
``` r
m3 <- lmerTest::lmer(Petal.Length ~ Sepal.Length + (1 | Species), data = iris)

coef(summary(m3, ddf = "lme4"))
#>               Estimate Std. Error     t value
#> (Intercept)  0.0525246 0.95576245  0.05495571
#> Sepal.Length 0.6341373 0.04524917 14.01433934
coef(summary(m3, ddf = "Satterthwaite"))
#>               Estimate Std. Error         df     t value     Pr(>|t|)
#> (Intercept)  0.0525246 0.95576245   2.339927  0.05495571 9.605145e-01
#> Sepal.Length 0.6341373 0.04524917 146.286388 14.01433934 7.882014e-29
coef(summary(m3, ddf = "Kenward-Roger"))
#>               Estimate Std. Error         df     t value     Pr(>|t|)
#> (Intercept)  0.0525246 0.95583828   2.342223  0.05495135 9.605137e-01
#> Sepal.Length 0.6341373 0.04529607 146.288076 13.99983109 8.598429e-29
```

(As well as the robust methods, I guess)

So this is only a matter of documenting - which is already covered in, say `?parameters::se_kenward`.

**Comment by bwiernik** [comment]
So what change are you proposing? Keep the robust se arguments separate?

**Comment by mattansb** [comment]
> Ideally, profile and uniroot would use LRT versus a model with the term removed (set to zero) to get p values, not Wald.

There would be several potential problems with such an approach:
1. That isn't always a 1:1 correspondence between parameters and terms.
2. Interactions... We'd be getting into type 2 vs type 3 ANOVA table-like stuff... No one wants that!

Why aren't Wald good enough here? The SE is based on the likelihood function (~profile etc)... So wouldn't p-values correspond nicely?

**Comment by bwiernik** [comment]
Don't over think it with (1) and (2) there. Profiling the likelihood along a parameter is fixing that parameter to some value, re-optimizing the parameters with that one fixed, and calculating the log likelihood for the updated model. 

Typically profiling is used to construct intervals by finding the point where difference in -2LL is 3.84/etc. 

A LRT for a single parameter is a likelihood profile, only we just fix the parameter to one value (0) instead of a range of values.

For a symmetrical likelihood function, Wald p values will be close, but, eg, for corrections and variance components, they won't match up well.

**Comment by bwiernik** [comment]
Yeah, that's why I said "we can ignore that for now".

**Comment by strengejacke** [comment]
> So what change are you proposing? Keep the robust se arguments separate?

I would suggest to only keep those `se_*()` methods where the standard error actually depends on the df-method and really changes. I think this is only the case for kenward-roger.

Furthermore, robust estimation affects the vcov-matrix, and thus standard error change. The question is, does it make (statistically) sense to combine robust estimation with different methods of df-approximation?

**Comment by mattansb** [comment]
@strengejacke Is this expected in the next release?

**Comment by strengejacke** [comment]
Yes, I wanted to have these changes in the next _parameters_ release.

**Comment by bwiernik** [comment]
Yes, it's fine to combine Huber and Satterthwaite/Rogers standard errors

**Comment by strengejacke** [comment]
What would you call the "default" CI method for classical `lm`? "wald"?

**Comment by mattansb** [comment]
That would be "residual", as it used a t distribution with the residuals df.

**Comment by mattansb** [comment]
Wald gives a z test (or t with df=Inf)

**Comment by mattansb** [comment]
In fact...

``` r
library(parameters)

m <- lme4::lmer(Reaction ~ Days + (Days | Subject), lme4::sleepstudy)
```

This should give a z-value, not a t-value with residual df.
```R
model_parameters(m, ci_method = "wald") 
#> # Fixed Effects
#> 
#> Parameter   | Coefficient |   SE |           95% CI | t(174) |      p
#> ---------------------------------------------------------------------
#> (Intercept) |      251.41 | 6.82 | [238.03, 264.78] |  36.84 | < .001
#> Days        |       10.47 | 1.55 | [  7.44,  13.50] |   6.77 | < .001
#> 
#> # Random Effects
#> 
#> Parameter               | Coefficient
#> -------------------------------------
#> SD (Intercept: Subject) |       24.74
#> SD (Days: Subject)      |        5.92
#> Cor (Intercept~Subject) |        0.07
#> SD (Residual)           |       25.59
#> 
#> Degrees of freedom based on Wald approximation.
```

<sup>Created on 2021-09-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by bwiernik** [comment]
I would call it a `"pivot"` or just `"t"` or maybe `"wald_t"`. Pivot would be my choice. (ncp intervals are also pivot intervals.)

The t-distribution based method is technically a Wald interval. "Wald" could refer to t/F tests/intervals, as well as to z/chi-square tests/intervals. The defining feature is that the test statistic is constructed as a precision-weighted distance between the estimated parameter value and the null value. That distance is asymptotically chi-squared distributed. 

For most models, the finite-sample distribution of that statistic is unknown, but for normal regression models, the finite sample properties are well-defined to follow t/F distributions.

**Comment by bwiernik** [comment]
Rather than "Wald" maybe we should just say "normal approximation"

**Comment by strengejacke** [comment]
I will get back to this issue the next days. We should probably...

1. Update the table from https://github.com/easystats/parameters/issues/570#issuecomment-889716268, where we also have a column "argument type", that defines the value for `ci_method`. I would suggest keeping it as simple and intuitive as possible, so that also people who are less familiar with sophisticated statistical jargon know what they request.
2. We need to revise the docs. In particular, it would be helpful having something similar that resembles @mattansb 's table as a section in the docs. The place would be https://github.com/easystats/parameters/blob/f44ae4b9d38600570d52e1acce84bdf2326427e2/R/1_model_parameters.R#L57 - is there anyone who is good at English who can write a first draft? <sub>somehow, I have @bwiernik in mind, not sure why... ;-)</sub>
3. Then I would revise the API accordingly, to have it consistent based on our decisions we made here.

**Comment by bwiernik** [comment]
@strengejacke Do we have a function (or could you make one) that could take a model, get its formula, drop a predictor term, and then return a working updated formula that could be passed to `update()`? Essentially I'd like to be able to automate this:

``` r
m2 <- glmmTMB::glmmTMB(mpg ~ drat + hp + drat:hp + (1 | cyl), data = mtcars)
fs <- list(
  "(Intercept)" = mpg ~ 0 + drat + hp + drat:hp + (1 | cyl),
  drat = mpg ~ hp + drat:hp + (1 | cyl),
  hp = mpg ~ drat + drat:hp + (1 | cyl),
  "drat:hp" = mpg ~ drat + hp + (1 | cyl)
)
us <- lapply(fs, function(x) {
  update(m2, formula = x)
})
ps <- do.call(rbind, lapply(us, function(x) as.data.frame(anova(m2, x)[2, 6:8])))
ps
#>                 Chisq Chi Df Pr(>Chisq)
#> (Intercept) 0.2777907      1 0.59815294
#> drat        3.9834670      1 0.04594889
#> hp          0.0377553      1 0.84593533
#> drat:hp   0.3132189      1 0.57571160
```

**Comment by mattansb** [comment]
@bwiernik isn't that a type 3 ANOVA, a-la `car::Anova ()`?

Also, you might be interested in https://github.com/singmann/monet

**Comment by mattansb** [comment]
``` r
m2 <- glmmTMB::glmmTMB(mpg ~ drat + hp + drat:hp + (1 | cyl), data = mtcars)
#> Warning in checkMatrixPackageVersion(): Package version inconsistency detected.
#> TMB was built with Matrix version 1.3.3
#> Current Matrix version is 1.3.4
#> Please re-install 'TMB' from source using install.packages('TMB', type = 'source') or ask CRAN for a binary version of 'TMB' matching CRAN's 'Matrix' package

car::Anova(m2, type = 3)
#> Analysis of Deviance Table (Type III Wald chisquare tests)
#> 
#> Response: mpg
#>              Chisq Df Pr(>Chisq)  
#> (Intercept) 0.2790  1    0.59736  
#> drat        5.0856  1    0.02413 *
#> hp          0.0378  1    0.84589  
#> drat:hp     0.3148  1    0.57477  
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

<sup>Created on 2021-09-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by bwiernik** [comment]
No, I'm looking to do likelihood ratio tests, not Wald tests.

**Comment by bwiernik** [comment]
Nvm, I think I figured it out after digging into the innards of `car::Anova.glm()`

**Comment by bwiernik** [comment]
Curiosity got me looking into implementing likelihood ratio based p values. Here is an example--they are virtually identical to normal wald p values for normal likelihood models, but somewhat different for other distributions:

``` r
profile_p <- function(m) {
  mm <- insight::get_modelmatrix(m)
  if (colnames(mm)[1] == "(Intercept)") {
    colnames(mm)[1] <- "Intercept"
  }
  mf <- insight::get_data(m)
  m_formula <- insight::find_formula(m)
  if ("random" %in% names(m_formula)) {
    m_formula[["random"]] <- stringr::str_replace(
      format(m_formula[["random"]]),
      "~", "")
  }
  mll <- logLik(m)[1]
  new_dat <- cbind(mf, mm)
  params <- colnames(mm)
  res <- data.frame(
    Parameter = params,
    Chi2 = NA,
    df = 1,
    p = NA
  )
  row.names(res) <- res$Parameter
  for (i in seq_along(params)) {
    new_formula <- m_formula
    new_cond <- as.character(stats::update.formula(
      m_formula$conditional,
      paste(". ~ 0 +", paste0("`", params[-i], "`", collapse = " + "))
    ))
    new_cond <- paste(new_cond[c(2, 1, 3)], collapse = " ")
    new_formula$conditional <- new_cond
    new_formula <- format(new_formula)
    new_mod <- update(
      m, 
      formula = new_formula, 
      data = new_dat
    )
    res[params[i], "Chi2"] <- -2 * (logLik(new_mod)[1] - mll)
  }
  res$p <- pchisq(res$Chi2, 1, lower.tail = FALSE)
  row.names(res) <- seq_len(nrow(res))
  return(res)
}

m2 <- glmmTMB::glmmTMB(count~spp * mined + (1|site), family = "poisson", data = glmmTMB::Salamanders)

lr_p <- profile_p(m2)
lr_p$p <- insight::format_p(lr_p$p)
lr_p
#>           Parameter       Chi2 df         p
#> 1         Intercept 45.2265007  1  p < .001
#> 2             sppPR  1.3282863  1 p = 0.249
#> 3             sppDM 15.9035091  1  p < .001
#> 4           sppEC-A  0.6795962  1 p = 0.410
#> 5           sppEC-L  6.8594281  1 p = 0.009
#> 6          sppDES-L 23.1711370  1  p < .001
#> 7             sppDF 23.1711370  1  p < .001
#> 8           minedno 44.3374584  1  p < .001
#> 9     sppPR:minedno  9.2680885  1 p = 0.002
#> 10    sppDM:minedno 13.2261261  1  p < .001
#> 11  sppEC-A:minedno  3.2236329  1 p = 0.073
#> 12  sppEC-L:minedno  2.5990309  1 p = 0.107
#> 13 sppDES-L:minedno 11.3619454  1  p < .001
#> 14    sppDF:minedno 23.8017664  1  p < .001

parameters::parameters(m2)
#> # Fixed Effects
#> 
#> Parameter                | Log-Mean |   SE |         95% CI |     z |      p
#> ----------------------------------------------------------------------------
#> (Intercept)              |    -3.38 | 0.73 | [-4.82, -1.94] | -4.60 | < .001
#> spp [PR]                 |     0.92 | 0.84 | [-0.72,  2.56] |  1.10 | 0.273 
#> spp [DM]                 |     2.25 | 0.74 | [ 0.79,  3.71] |  3.03 | 0.002 
#> spp [EC-A]               |     0.69 | 0.87 | [-1.00,  2.39] |  0.80 | 0.423 
#> spp [EC-L]               |     1.70 | 0.77 | [ 0.20,  3.21] |  2.22 | 0.027 
#> spp [DES-L]              |     2.53 | 0.73 | [ 1.09,  3.97] |  3.44 | < .001
#> spp [DF]                 |     2.53 | 0.73 | [ 1.09,  3.97] |  3.44 | < .001
#> mined [no]               |     4.11 | 0.76 | [ 2.62,  5.60] |  5.42 | < .001
#> spp [PR] * mined [no]    |    -2.49 | 0.87 | [-4.19, -0.79] | -2.86 | 0.004 
#> spp [DM] * mined [no]    |    -2.15 | 0.76 | [-3.63, -0.67] | -2.85 | 0.004 
#> spp [EC-A] * mined [no]  |    -1.53 | 0.88 | [-3.26,  0.20] | -1.73 | 0.084 
#> spp [EC-L] * mined [no]  |    -1.12 | 0.78 | [-2.65,  0.40] | -1.44 | 0.150 
#> spp [DES-L] * mined [no] |    -1.95 | 0.74 | [-3.41, -0.49] | -2.62 | 0.009 
#> spp [DF] * mined [no]    |    -2.67 | 0.75 | [-4.13, -1.20] | -3.56 | < .001
#> 
#> # Random Effects
#> 
#> Parameter            | Coefficient
#> ----------------------------------
#> SD (Intercept: site) |        0.58
#> SD (Residual)        |        1.00
```

<sup>Created on 2021-09-24 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by mattansb** [comment]
Cool!
We could generalize this to a function that:
1. Refits subset models by removing a (fixed effect) parameter *or* term
2. Computes an arbitrary function(s) on the original and each subset model. `function(m_orig, m_subset) ...`

This way we can get for each parameter/term  LRT, semi-partial R2, and more

**Comment by mattansb** [comment]
``` r
compare_nested <- function(m, fn, by = c("term", "parameter")) {
  by <- match.arg(by)
  
  mf <- insight::get_data(m)
  m_formula <- insight::find_formula(m)
  mm <- insight::get_modelmatrix(m)
  
  if ("random" %in% names(m_formula)) {
    m_formula[["random"]] <- gsub("~", "", format(m_formula[["random"]]))
  }
  
  if (colnames(mm)[1] == "(Intercept)") {
    colnames(mm)[1] <- "Intercept"
  }
  new_dat <- cbind(mf, mm)
  
  if (by == "term") {
    v_assign <- attr(mm, "assign")
    k_assign <- unique(v_assign)
    subset_formula <- vector("list", length(k_assign))
    
    for (k in seq_along(k_assign)) {
      tmp_formula <- m_formula
      tmp_cond <- stats::update.formula(
        tmp_formula$conditional,
        paste(". ~ 0 +", paste0("`", colnames(mm)[v_assign!=k_assign[k]], "`", collapse = " + "))
      )
      tmp_formula$conditional <- paste(as.character(tmp_cond)[c(2, 1, 3)], collapse = " ")
      subset_formula[k] <- format(tmp_formula)
    }
    # names(subset_formula) <- # Terms
  } else {
    subset_formula <- vector("list", ncol(mm))
    v_pars <- colnames(mm)
    
    for (k in seq_along(v_pars)) {
      tmp_formula <- m_formula
      tmp_cond <- stats::update.formula(
        tmp_formula$conditional,
        paste(". ~ 0 +", paste0("`", v_pars[-k], "`", collapse = " + "))
      )
      tmp_formula$conditional <- paste(as.character(tmp_cond)[c(2, 1, 3)], collapse = " ")
      subset_formula[k] <- format(tmp_formula)
    }
    names(subset_formula) <- v_pars
  }
  
  
  res <- lapply(subset_formula, function(sf) {
    s_mod <- update(m, formula = sf, data = new_dat)
    fn(m, s_mod)
  })
  return(res)
}

m2 <- glmmTMB::glmmTMB(count~spp * mined + (1|site), family = "poisson", data = glmmTMB::Salamanders)
#> Warning in checkMatrixPackageVersion(): Package version inconsistency detected.
#> TMB was built with Matrix version 1.3.3
#> Current Matrix version is 1.3.4
#> Please re-install 'TMB' from source using install.packages('TMB', type = 'source') or ask CRAN for a binary version of 'TMB' matching CRAN's 'Matrix' package

do_LRT <- function(m, sub_m) {
  ll1 <- logLik(m)
  ll2 <- logLik(sub_m)
  
  chisq <- -2 * (ll2[1] - ll1[1])
  df <- attr(ll1, "df") - attr(ll2, "df")
  p <- pchisq(chisq, df, lower.tail = FALSE)
  
  data.frame(chisq, df, p)
}

compare_nested(m2, do_LRT) |> do.call(what = "rbind")
#>      chisq df            p
#> 1 45.22650  1 1.755139e-11
#> 2 49.66650  6 5.483212e-09
#> 3 44.33746  1 2.763769e-11
#> 4 34.55771  6 5.248732e-06
compare_nested(m2, do_LRT, by = "par") |> do.call(what = "rbind")
#>                       chisq df            p
#> Intercept        45.2265007  1 1.755139e-11
#> sppPR             1.3282863  1 2.491103e-01
#> sppDM            15.9035091  1 6.665502e-05
#> sppEC-A           0.6795962  1 4.097258e-01
#> sppEC-L           6.8594281  1 8.817476e-03
#> sppDES-L         23.1711370  1 1.482052e-06
#> sppDF            23.1711370  1 1.482052e-06
#> minedno          44.3374584  1 2.763769e-11
#> sppPR:minedno     9.2680885  1 2.331809e-03
#> sppDM:minedno    13.2261261  1 2.760738e-04
#> sppEC-A:minedno   3.2236329  1 7.258238e-02
#> sppEC-L:minedno   2.5990309  1 1.069291e-01
#> sppDES-L:minedno 11.3619454  1 7.496426e-04
#> sppDF:minedno    23.8017664  1 1.067844e-06
```

<sup>Created on 2021-09-25 by the [reprex package](https://reprex.tidyverse.org) (v2.0.1)</sup>

**Comment by strengejacke** [comment]
Ehm, your discussion doesn't really help me getting a clearer picture of how to implement this. 😬

**Comment by bwiernik** [comment]
> This way we can get for each parameter/term LRT, semi-partial R2, and more

That's a great idea

> Ehm, your discussion doesn't really help me getting a clearer picture of how to implement this. 😬

We can take it outside

**Comment by mattansb** [comment]
Oh, I just realized I can add the term names like the parameter names. Will fix this when I get home.

**Comment by bwiernik** [comment]
@mattansb Could you also allow `fn` to be a list of functions, so that we could compute multiple comparison results at once?

**Comment by mattansb** [comment]
Sure.
Should the output be coerced to a data frame (cbind functions, rbind terms/parameters)? Which would require the fns to always return data frames?
Or should we let it be flexible and return a list of lists?

**Comment by bwiernik** [comment]
I think return a list

**Comment by bwiernik** [comment]
@strengejacke How does this look?

| `ci_method` | Applicable models | CI computation | p value computation | Changed behavior? |
| -------------- | -------------- | -------------- | -------------- | -------------- |
| `hdi` | All models | For models with iterations, highest density interval; for others, wald | pd | x |
| `eti` | All models | For models with iterations, highest density interval; for others, wald | pd | x |
| `si` | Bayesian models | Support interval | pd |  |
| `bci` | Bayesian/bootstrap models | BCa interval | pd |  |
| `bcai` | Bayesian/bootstrap models | BCa interval | pd |  |
| `wald` | Non-Bayesian models | For normal linear fixed effects models, t distribution with residual df; for others, normal distribution | wald p (t or normal) | some bugs? |
| `profile` | Non-Bayesian models | Profile the likelihood curve for a paramter, use linear interpolation to find where likelihood ratio equals critical value | wald p (normal) |  |
| `uniroot` | TMB models | Profile the likelihood curve for a paramter, use root finding to find where likelihood ratio equals critical value | wald p (normal) |  |
| `profile_p` | Non-Bayesian models | Same as profile | likelihood ratio test | new |
| `uniroot_p` | TMB models | Same as unitroot | likelihood ratio test | new |
| `ml1` | LMMs | t distribution with ml1 df | wald p (t) |  |
| `betwithin` | LMMs | t distribution with between-within df | wald p (t) |  |
| `satterthwaite` | LMMs | t distribution with Satterthwaite df | wald p (t) |  |
| `kenward` | LMMs | t distribution with Kenward SE, df | wald p (t) |  |
| `pivot` | Non-Bayesian models | Same as wald | wald p (t or normal) | new |
| `normal` | All models | normal distribution | wald p (normal) | bug |
| `residual` | Non-Bayesian models | t distribution with residual df | wald p (t) |  |


### Notes
We don't currently support other bootstrap types (normal approximation, basic [bootstrap t], studentized).
For Bayesian and bootstrap models, we could support wald/normal/residual/etc. using the estimated SE from the distribution.
For uniroot, we could fall back to profile for other types of models.
We don't currently support other methods that might be available for some classes (e.g., `metafor::rma.uni`).

**Comment by bwiernik** [comment]
With respect to `robust`, I looked into it more, and I don't see any work, either theoretical or simulation, combining any sandwich estimators with Satterthwaite or Kenward-Roger approximations. [SAS permits combining sandwich estimators](https://support.sas.com/documentation/cdl/en/statug/63033/HTML/default/viewer.htm#statug_mixed_sect015.htm#statug.mixed.mixedmodelddfm) with the residual and between-within df methods, as well as their default "containment" method, but not SW or KR.

So I agree with our existing approach to keep `robust` a separate argument. `robust = TRUE` should be combinable to with `c("wald", "ml1", "betwithin", "pivot", "normal", and "residual")`.

**Comment by bwiernik** [comment]
I can put the table I made into docs this weekend

**Comment by strengejacke** [comment]
What still needs to be done?

- The documentation of the `method` resp. `ci_method` argument in `ci()` / `p_value()` and `model_parameters()`.
- For some models, not all `ci_method` options are available. Currently, sometimes this is silently changed into `"wald"`. We might be more verbose here and do some stricter checking of arguments, and tell the user if s/he used invalid options. But I think if we have a proper documentation, it should be clear what can be used.
- Some of the new options shown in https://github.com/easystats/parameters/issues/570#issuecomment-927138081 are not yet implemented.
- We need to check the additional info message about uncertainty intervals and p-values printed from `model_parameters()`, to see if everything is correct.

**Comment by strengejacke** [comment]
@bwiernik The "Between-within denominator degrees of freedom approximation" is also applicable for GLMMs (see https://doi.org/10.1186/s12874-015-0026-x), so not only t-distribution.

**Comment by mattansb** [comment]
@strengejacke I assume that `p_value / ci` with `method = "profile" / "boot"` don't compute the CIs and p-value (based on `estimate +/- se * qt(alpha, df)` and `2 * pt(abs(estimate / se), df, lower.tail = FALSE)`, correct? But I couldn't find in the new code where that happens.

Also, for any transformations (e.g., `exponentiate = TRUE`) happen *after* p/ci computations?

**Comment by mattansb** [comment]
- [ ] `si` should *not* be a ci method
(Definitely not for non-Bayesian models).
It gives information about the relative change in likelihood (so requires priors), not about the posterior.

**Comment by strengejacke** [comment]
@vincentarelbundock pinging you to make you aware of these changes (see also section "Confidence intervals and approximation of degrees of freedom" in https://easystats.github.io/parameters/reference/model_parameters.html).
